### DISCUSSION TITLE
In workflows, how to use ghcr docker image?

### DISCUSSION THREAD

**Discussion by punkbit** [community/community#25684] [Actions]
<p>Hi,</p>
<p>After updating my docker image and pushing to ghcr, the changes are not available when running the workflow!</p>
<p>The workflow looks like:</p>
<pre><code class="lang-auto">name: Some name

on:
  pull_request:
    branches: [ main ]

  workflow_dispatch:

jobs:
  build:
    runs-on: ubuntu-latest
    container:
      image: ghcr.io/organisation/docker-image:tag
      credentials:
        username: ${{ github.actor }}
        password: ${{ secrets.PRIVATE_ACCESS_TOKEN }}
</code></pre>
<p>The workdir in the docker image is:</p>
<pre><code class="lang-auto">/github/home
</code></pre>
<p>I’m still trying to figure out what’s going on, to the point where I stopped relying in the <code>latest</code> tag and used a very specific tag that contains the changes I want. To b more specific, there’s a binary that I added and is available when I check it in my local machine (docker run), but it’s not when running the cloud (github workflow).</p>
<hr>
<p>Did a lot of tests and even uploading the same image under a different name, to avoid any caching, but same issue.</p>
<p>Locally, when I run the image, I can instantiate it and open a shell and see the binaries etc, but not once in the cloud.</p>

**Comment by airtower-luna** [comment]
<aside class="quote no-group" data-username="heldrida" data-post="1" data-topic="190249">
<div class="title">
<div class="quote-controls"></div>
 heldrida:</div>
<blockquote>
<p>After updating my docker image and pushing to ghcr, the changes are not available when running the workflow!</p>
</blockquote>
</aside>
<p>Could you show us exactly how the update and push happens? The way the image is used looks correct (assuming the credentials are), so it’s probably something there.</p>

**Comment by punkbit** [comment]
<p>The Dockerfile is pushed manually to mitigate any issues:</p>
<pre><code class="lang-auto">docker build -t image:latest

docker tag hash ghcr.io/org/image:latest

docker push ghcr.io/org/image:latest
</code></pre>
<p>The project requires Rust, Rustup, Cargo, etc. Which is all in the Dockerfile, as tested locally.</p>
<p>For some reason, <code>rustup</code> should be located in <code>/github/home/.cargo</code>, instead of the default cargo installation directory which is <code>/root/.cargo/bin</code>.</p>
<p>This step is done in the Dockerfile, as:</p>
<pre><code class="lang-auto">ln -s "/root/.cargo/bin" /github/home/.cargo
</code></pre>
<p>Which then, when I do open a shell session locally to the Docker instance, I confirm everything is ok and operates as expected:</p>
<pre><code class="lang-auto">docker run -v $(pwd):/github/workdir/ -it ghcr.io/org/image:latest /bin/bash
</code></pre>
<p>Lists the expected files or symlink when:</p>
<pre><code class="lang-auto">ln -ls /github/home
</code></pre>
<p>Now, in the cloud, if I do the same as done locally:</p>
<pre><code class="lang-auto">jobs:
  build:
    runs-on: ubuntu-latest
    container:
      image: ghcr.io/org/image:latest
      credentials:
        username: ${{ github.actor }}
        password: ${{ secrets.PRIVATE_ACCESS_TOKEN }}

    steps:
      - name: DEBUG
        run: |
          ls -la /github/home
</code></pre>
<p>The output is empty or does not have the <code>.cargo</code> symlink!</p>
<pre><code class="lang-auto">.
..
</code></pre>
<p>I understand I can have the <code>Rust Toolchain</code> github action, but if I go with this option, I need to build Rust source code which is extremely slow. I also understand I can use cache by using key names and validation but I just want to use the Docker instance as I have tested locally and not have to include or build certain dependencies.</p>
<pre><code class="lang-auto">      - name: Rust toolchain
        uses: actions-rs/toolchain@v1
        with:
          toolchain: stable
          target: wasm32-unknown-unknown
</code></pre>

**Comment by punkbit** [comment]
<p>Just thought about something that I need to test.</p>
<p>The path I’ve been using in the Docker file is:</p>
<pre><code class="lang-auto">ENV PATH="/root/.cargo/bin:${PATH}"
</code></pre>
<p>But I’ll now try:</p>
<pre><code class="lang-auto">export PATH="$HOME/.cargo/bin:$PATH"
</code></pre>
<p>And also:</p>
<pre><code class="lang-auto">ln -s "/root/.cargo/bin" "$HOME/.cargo"
</code></pre>
<hr>
<p>Actually this wouldn’t work as effectively, $HOME throws, when running locally:</p>
<pre><code class="lang-auto">failed to create symbolic link '/root/.cargo/bin': File exists
</code></pre>
<p>So, I guess the original should be alright:</p>
<pre><code class="lang-auto">RUN ln -s "/root/.cargo/bin" /github/home/.cargo
</code></pre>
<p>And probably this <code>/github/home</code> location is a volume that is mounted automatically maybe?! And so anything I do in advance is effectively removed?! I do remember testing doing the symlink as a step in the github action workflow, but can double check.</p>

**Comment by airtower-luna** [comment, marked as answer]
<aside class="quote no-group" data-username="heldrida" data-post="3" data-topic="190249">
<div class="title">
<div class="quote-controls"></div>
 heldrida:</div>
<blockquote>
<p>For some reason, <code>rustup</code> should be located in <code>/github/home/.cargo</code> , instead of the default cargo installation directory which is <code>/root/.cargo/bin</code> .</p>
</blockquote>
</aside>
<p>This right there is the problem. When using a container as a job container, the runner mounts several directories from the host into the container. <code>/github/home</code> is one of the target directories, which hides anything you might have put there in the container image (check the “Initialize containers” step in your workflow logs).</p>
<p>So if you want your software to be available inside the job container, it really needs to go <em>anywhere but</em> <code>/github/</code> inside the container image. :slightly_smiling_face:</p>

**Comment by punkbit** [comment]
<p>Thanks! That’s good to know, hope it helps other users.</p>
<p>I have also tried the symlink in the Github Action workflow, but I will double check (what I meant is, during some testing, for alternative solutions). If do it as a step in the workflow, do you think it’ll work?</p>

**Comment by punkbit** [comment]
<aside class="quote no-group" data-username="airtower-luna" data-post="5" data-topic="190249">
<div class="title">
<div class="quote-controls"></div>
 airtower-luna:</div>
<blockquote>
<p>Initialize containers</p>
</blockquote>
</aside>
<p>Do you know when the doc for “Initialize containers” is located? trying to find it</p>

**Comment by airtower-luna** [comment]
<p>I don’t think there is a specific documentation. I mean you should check the workflow logs, where you can see the exact commands the runner is using. Here’s an example from one of my workflows:</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/3/8/382e85e896dd3902040d24b1d2d36deb9f740bf1.png" data-download-href="/uploads/short-url/810pDOSdCIa9AhyJOqfp5hCkxRT.png?dl=1" title="grafik"><img src="https://user-images.githubusercontent.com/5814919/181092494-59bafcd3-b26d-4697-9ade-144d7a967ce6.png" alt="grafik" data-base62-sha1="810pDOSdCIa9AhyJOqfp5hCkxRT" width="546" height="500" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/3/8/382e85e896dd3902040d24b1d2d36deb9f740bf1_2_546x500.png, https://aws1.discourse-cdn.com/github/original/3X/3/8/382e85e896dd3902040d24b1d2d36deb9f740bf1.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/3/8/382e85e896dd3902040d24b1d2d36deb9f740bf1.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/3/8/382e85e896dd3902040d24b1d2d36deb9f740bf1_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">grafik</span><span class="informations">794×726 108 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>

**Comment by punkbit** [comment]
<p>Thank you for letting me know!</p>

**Comment by lirantal** [comment]
<p>What a gem, I didn’t know that either!</p>
